src/collisions.rs

```diff
@@ -0,0 +1,462 @@
+    }
+
+    pub fn center(&self) -> Vec2 {
+        Vec2::new(self.x + self.width / 2.0, self.y + self.height / 2.0)
+    }
+
+    pub fn contains(&self, point: Vec2) -> bool {
+        self.x <= point.x
+            && point.x < (self.x + self.width)
+            && self.y <= point.y
+            && point.y < (self.y + self.height)
+    }
+
+    pub fn inflate(&mut self, horizontal: f32, vertical: f32) {
+        self.x -= horizontal;
+        self.y -= vertical;
+        self.width += horizontal * 2.0;
+        self.height += vertical * 2.0;
+    }
+
+    pub fn intersects(&self, other: Rect) -> bool {
+        other.left() < self.right()
+            && self.left() < other.right()
+            && other.top() < self.bottom()
+            && self.top() < other.bottom()
+    }
+
+    pub fn union(&self, other: &Rect) -> Rect {
+        let x = self.x.min(other.x);
+        let y = self.y.min(other.y);
+
+        Rect::new(
+            x,
+            y,
+            self.right().max(other.right()) - x,
+            self.bottom().max(other.bottom()) - y,
+        )
+    }
+
+    pub fn union_vec2(&self, vec: &Vec2) -> Rect {
+        let rect = Rect::new(vec.x, vec.y, 0.0, 0.0);
+        self.union(&rect)
+    }
+
+    pub fn closest_point_to_origin(&self) -> Vec2 {
+        let max = self.max();
+        let mut min_dist = self.x.abs();
+        let mut bounds_point = Vec2::new(self.x, 0.0);
+
+        if max.x.abs() < min_dist {
+            min_dist = max.x.abs();
+            bounds_point.x = max.x;
+            bounds_point.y = 0.0;
+        }
+
+        if max.y.abs() < min_dist {
+            min_dist = max.y.abs();
+            bounds_point.x = 0.0;
+            bounds_point.y = max.y;
+        }
+
+        if self.y.abs() < min_dist {
+            // min_dist = self.y.abs();
+            bounds_point.x = 0.0;
+            bounds_point.y = self.y;
+        }
+
+        bounds_point
+    }
+
+    /// Returns (Closest, EdgeNormal)
+    pub fn closest_point_on_border(&self, point: Vec2) -> (Vec2, Vec2) {
+        let mut edge_normal = Vec2::ZERO;
+
+        let mut res = Vec2::new(0.0, 0.0);
+        res.x = point.x.clamp(self.left(), self.right());
+        res.y = point.y.clamp(self.top(), self.bottom());
+
+        if self.contains(res) {
+            let dl = res.x - self.left();
+            let dr = self.right() - res.x;
+            let dt = res.y - self.top();
+            let db = self.bottom() - res.y;
+
+            let min = dl.min(dr.min(dt.min(db)));
+            if min == dt {
+                res.y = self.top();
+                edge_normal.y = -1.0;
+            } else if min == db {
+                res.y = self.bottom();
+                edge_normal.y = 1.0;
+            } else if min == dl {
+                res.x = self.left();
+                edge_normal.x = -1.0;
+            } else {
+                res.x = self.right();
+                edge_normal.x = 1.0;
+            }
+        } else {
```

Comment:
maybe we should remove unneeded else so there will and just return our values in the upper if and do downward logic without else brackets?

Reply:
after if make return, remove else. else is unneeded in this case imo

Reply:
```
if ... {
   ...
   ...
   return ...
}
  ...
  ...
  return ...
  ```

Reply:
We need an `else` here

Reply:
why

Reply:
Fixed